frontend/public/components/marketplace/kubernetes-marketplace-items.jsx

```diff
@@ -0,0 +1,288 @@
+    }
+
+    // If only filters have changed, update filter items and filter counts
+    if (updatedState.filters || filters !== prevState.filters) {
+      Object.assign(updatedState, this.getFilterItemsAndCounts(updatedState.filters || filters));
+    }
+
+    this.setState(updatedState);
+  }
+
+  getFilterItemsAndCounts(filters) {
+    const { items } = this.props;
+
+    const itemsAndCounts = {
+      filteredItems: {},
+      filterCounts: {},
+    };
+
+    // Filter items by keyword first
+    const filteredByKeyword = this.filterByKeyword(items, filters);
+
+    // Apply each filter property individually. Example:
+    //  filteredByCategory = {
+    //    provider: [/*array of items filtered by provider*/],
+    //    healthIndex: [/*array of items filtered by healthIndex*/],
+    //  };
+    const filteredByCategory = this.filterByCategory(filteredByKeyword, filters);
+
+    // Intersection of individually applied filters is all filters
+    // In the case no filters are active, returns items filteredByKeyword
+    const filteredItems = _.intersection(..._.values(filteredByCategory), filteredByKeyword);
+    itemsAndCounts.filteredItems = _.sortBy(filteredItems, 'name');
+
+    // Get counts for filtered items in each category
+    itemsAndCounts.filterCounts = this.getFilterCountsFromCategories(filteredByKeyword, filters, filteredByCategory);
+
+    return itemsAndCounts;
+  }
+
+  filterByCategory(items, filters) {
+    // Filter items by each filter category
+    return _.reduce(filters, (filtered, category, key) => {
+      if (key === 'keyword') {
+        return filtered;
+      }
+
+      // Only apply active filters
+      const activeFilters = _.filter(category, 'active');
+      if (activeFilters.length) {
+        const values = _.map(activeFilters, 'value');
+        filtered[key] = _.filter(items, item => values.includes(item[key]));
+      }
+
+      return filtered;
+    }, {});
+  }
+
+  filterByKeyword(items, filters) {
+    const { keyword } = filters;
+    if (keyword.active) {
+      const filterString = keyword.value.toLowerCase();
+      return _.filter(items, item => item.name.toLowerCase().includes(filterString) ||
+        (item.description && item.description.toLowerCase().includes(filterString)) ||
+        (item.tags && item.tags.includes(filterString)));
+    }
+    return items;
+  }
+
+  getFilterCountsFromCategories(items, filters, itemsFilteredByCategory) {
+    // Input items, filters, and mapping from filter property to filtered items
+    // Returns object mapping filter property to item counts
+    return _.reduce(filters, (counts, category, key) => {
+      if (key === 'keyword') {
+        return counts;
+      }
+
+      // Apply filters for all properties except for current key property
+      const itemsByCategory = _.filter(itemsFilteredByCategory, (categoryItems, categoryKey) => categoryKey !== key);
+      const count = _.countBy(_.intersection(...itemsByCategory, items), key);
```

Comment:
Is `items` all items? Then isn't the intersection of `itemsByCategory` and `items` going to be the same as `itemsByCategory`?

Reply:
In other words, is this the same as...

```suggestion
      const count = _.countBy(itemsByCategory, key);
```

Reply:
`itemsByCategory` is an array of arrays. I'm finding the intersection on the inner arrays to figure out the items to use for counts in a given filter group. I've added the `items` array for good measure since itemsByCategory may be an empty array if no filters are active.